lib/uplink/ext/test_utils.go

```diff
@@ -0,0 +1,180 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package main
+
+// #cgo CFLAGS: -g -Wall
+// #include <stdlib.h>
+// #ifndef STORJ_HEADERS
+//   #define STORJ_HEADERS
+//   #include "c/headers/main.h"
+// #endif
+import "C"
+import (
+	"context"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"testing"
+	"unsafe"
+
+	"github.com/skyrings/skyring-common/tools/uuid"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap/zaptest"
+
+	"storj.io/storj/internal/testcontext"
+	"storj.io/storj/internal/testplanet"
+	"storj.io/storj/lib/uplink"
+	"storj.io/storj/pkg/macaroon"
+	"storj.io/storj/satellite/console"
+)
+
+/* C types */
+
+// Cchar is an alias to the c character pointer type.
+type Cchar = *C.char
+
+// CUint is an alias to the c uint type.
+type CUint = C.uint
+
+/* Ref types */
+
+// CAPIKeyRef is an alias to the APIKeyRef_t c type.
+type CAPIKeyRef = C.APIKeyRef_t
+
+var (
+	// CLibDir is the path to the storj/lib/uplink/ext/c directory.
+	CLibDir string
+	// CSrcDir is the path to the storj/lib/uplink/ext/c/src directory.
+	CSrcDir string
+	// CTestsDir is the path to the storj/lib/uplink/ext/c/src directory.
+	CTestsDir string
+	// LibuplinkSO is the path to the storj/lib/uplink/ext/uplink-cgo.so shared object.
+	LibuplinkSO string
+
+	testConfig = new(uplink.Config)
+)
+
+func init() {
+	// TODO: is there a cleaner way to do this?
```

Comment:
Yes, build the `.so` into a temporary directory and the working directory points to the package.

Reply:
Cgo needs the paths to included c files (in the `import C` comment) to be valid but so do the c tests when we build them. The problem is that the header file produced by cgo (uplink-cgo.h) includes those paths so you can't move the `.so` and it's header file relative to the package root without either breaking cgo compilation or the c test compilation

Reply:
Unless we rewrite the `uplink-cgo.h` header file to change the paths after it's written, I'm not sure how else we could move it.

Reply:
I'm not following.

As I understand the tests are fully in C, so cgo doesn't really matter here. Since you are compiling tests separately anyways you can set include paths as needed.